stream.go

```diff
@@ -239,75 +247,114 @@ func (s *Stream) sendReset() error {
 
 // Reset resets the stream (forcibly closes the stream)
 func (s *Stream) Reset() error {
+	sendReset := false
 	s.stateLock.Lock()
 	switch s.state {
-	case streamInit:
-		// No need to send anything.
-		s.state = streamReset
-		s.stateLock.Unlock()
-		return nil
-	case streamClosed, streamReset:
+	case streamFinished:
 		s.stateLock.Unlock()
 		return nil
+	case streamInit:
 	case streamSYNSent, streamSYNReceived, streamEstablished:
-	case streamLocalClose, streamRemoteClose:
+		sendReset = true
 	default:
 		panic("unhandled state")
 	}
-	s.state = streamReset
-	s.stateLock.Unlock()
 
-	err := s.sendReset()
+	// at least one direction is open, we need to reset.
+
+	// If we've already sent/received an EOF, no need to reset that side.
+	if s.writeState == halfOpen {
+		s.writeState = halfReset
+	}
+	if s.readState == halfOpen {
+		s.readState = halfReset
+	}
+	s.state = streamFinished
 	s.notifyWaiting()
+	s.stateLock.Unlock()
+	if sendReset {
+		_ = s.sendReset()
+	}
 	s.cleanup()
-
-	return err
+	return nil
 }
 
-// Close is used to close the stream
-func (s *Stream) Close() error {
-	closeStream := false
+// CloseWrite is used to close the stream for writing.
+func (s *Stream) CloseWrite() error {
 	s.stateLock.Lock()
-	switch s.state {
-	case streamInit, streamSYNSent, streamSYNReceived, streamEstablished:
-		s.state = streamLocalClose
-		goto SEND_CLOSE
+	switch s.writeState {
+	case halfOpen:
```

Comment:
I would add either a comment or a break here too, it's muddying the intentions to have it empty.

Reply:
Sure, but I don't see it. All of the other cases are aborts.

Reply:
Is this your C experience speaking?

Reply:
yeah, these things scare me :)

Reply:
Well, break in switch statements scare me because they tend to be _very_ confusing inside loops. I'd strongly prefer not to put them here.